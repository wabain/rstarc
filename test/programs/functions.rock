Put 1 into X
Put 1 into Y
Put 1 into Z

Func One takes X and Y
    Say "x"
    Say X
    Say "y"
    Say Y
    Say "z"
    Say Z
    Say ""
    Say "incrementing within function"
    If Y is mysterious
      Put 0 over 0 into Y

    Build X up
    Build Y up
    Build Z up
    Say "x"
    Say X
    Say "y"
    Say Y
    Say "z"
    Say Z
    Say ""


(
    Expectations:

    X is passed the value 1 as an argument, so it should be updated to
    2 within the function and left untouched outside.

    Y is not passed a value as an argument, so it should be given the
    value mysterious within the function and left untouched outside.

    Z is not a function argument so assigning to it within the function
    should affect the value outside.
)

Say "invoking function"
Put Func One taking 1 into Temp

Say "checking globals"
Say "x"
Say X
Say "y"
Say Y
Say "z"
Say Z
