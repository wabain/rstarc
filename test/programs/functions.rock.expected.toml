[tests]
enabled = ["run"]

["run.exec"]
returncode = "1"
stderr = """
error: using non-local variable 'Z' in compiled code is not supported.

Try executing your program using 'rstarc run --interpret <program>' instead

11:      Say Z
             ^
"""

["run.interpret"]
returncode = "0"
stdout = """
invoking function
x
1
y
mysterious
z
1

incrementing within function
x
2
y
NaN
z
2

checking globals
x
1
y
1
z
2
"""
